tests/GridFS/StreamWrapperFunctionalTest.php

```diff
@@ -204,4 +229,138 @@ public function testWritableStreamWrite(): void
 
         $this->assertSame(6, fwrite($stream, 'foobar'));
     }
+
+    /** @dataProvider provideUrl */
+    public function testStreamWithContextResolver(string $url, string $expectedFilename): void
+    {
+        $this->bucket->registerGlobalStreamWrapperAlias('bucket');
+
+        $stream = fopen($url, 'wb');
+
+        $this->assertSame(6, fwrite($stream, 'foobar'));
+        $this->assertTrue(fclose($stream));
+
+        $file = $this->filesCollection->findOne(['filename' => $expectedFilename]);
+        $this->assertNotNull($file);
+
+        $stream = fopen($url, 'rb');
+
+        $this->assertSame('foobar', fread($stream, 10));
+        $this->assertTrue(fclose($stream));
+    }
+
+    public static function provideUrl()
+    {
+        yield 'simple file' => ['gridfs://bucket/filename', 'filename'];
+        yield 'subdirectory file' => ['gridfs://bucket/path/to/filename.txt', 'path/to/filename.txt'];
+        yield 'question mark can be used in file name' => ['gridfs://bucket/file%20name?foo=bar', 'file%20name?foo=bar'];
+    }
+
+    public function testFilePutAndGetContents(): void
+    {
+        $this->bucket->registerGlobalStreamWrapperAlias('bucket');
+
+        $filename = 'gridfs://bucket/path/to/filename';
+
+        $this->assertSame(6, file_put_contents($filename, 'foobar'));
+
+        $file = $this->filesCollection->findOne(['filename' => 'path/to/filename']);
+        $this->assertNotNull($file);
+
+        $this->assertSame('foobar', file_get_contents($filename));
+    }
+
+    public function testEmptyFilename(): void
+    {
+        $this->bucket->registerGlobalStreamWrapperAlias('bucket');
+
+        $filename = 'gridfs://bucket';
+
+        $this->assertSame(6, file_put_contents($filename, 'foobar'));
+
+        $file = $this->filesCollection->findOne(['filename' => '']);
+        $this->assertNotNull($file);
+
+        $this->assertSame('foobar', file_get_contents($filename));
+    }
+
+    public function testOpenSpecificRevision(): void
+    {
+        $this->bucket->registerGlobalStreamWrapperAlias('bucket');
+
+        $filename = 'gridfs://bucket/path/to/filename';
+
+        // Insert 3 revisions, wait 1ms between each to ensure they have different uploadDate
+        file_put_contents($filename, 'version 0');
+        usleep(1000);
```

Comment:
@jmikola Added this 1ms delay to fix https://github.com/mongodb/mongo-php-library/pull/1138#issuecomment-1858364567